master/buildbot/steps/source/mtn.py

```diff
@@ -83,14 +82,21 @@ def checkInstall(monotoneInstalled):
                 raise BuildSlaveTooOldError("Monotone is not installed on slave")
             return 0
         d.addCallback(lambda _: self._checkDb())
-        d.addCallback(lambda _: self.sourcedirIsPatched())
+        d.addCallback(lambda _: self._retryPull())
 
-        @d.addCallback
-        def checkPatched(patched):
-            if patched:
-                return self.clean()
-            else:
-                return 0
+        # If we're not throwing away the workdir, check if it's
+        # somehow patched or modified and revert.
+        if self.mode != 'full' or self.method not in ('clobber', 'copy'):
+            d.addCallback(lambda _: self.sourcedirIsPatched())
+
+            @d.addCallback
```

Comment:
for big patches like this, we usually ask contributors to change the code to inlineCallbacks, which are so much easier to write and review